hazelcast/src/test/java/com/hazelcast/spi/addresslocator/AddressLocatorTest.java

```diff
@@ -0,0 +1,136 @@
+                .setEnabled(true)
+                .setClassname(AddressLocatorWithStaticProperties.class.getName())
+                .getProperties().setProperty("propName", "propValue");
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+
+        Hazelcast.newHazelcastInstance(config);
+        String property = AddressLocatorWithStaticProperties.properties.getProperty("propName");
+        assertEquals("propValue", property);
+    }
+
+    @Test
+    public void testSimpleAddressLocatorIsInjected() {
+        Config config = new Config();
+        config.getNetworkConfig().getAddressLocatorConfig()
+                .setEnabled(true)
+                .setClassname(SimpleAddressLocator.class.getName());
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        InetSocketAddress inetSocketAddress = (InetSocketAddress) instance.getLocalEndpoint().getSocketAddress();
+
+        assertEquals(inetSocketAddress.getPort(), 9999);
+    }
+
+    @Test(expected = ConfigurationException.class)
+    public void testFailFastWhenNoMatchingConstructorIsFound() {
+        Config config = new Config();
+        config.getNetworkConfig().getAddressLocatorConfig()
+                .setEnabled(true)
+                .setClassname(SimpleAddressLocator.class.getName())
+                .getProperties().setProperty("foo", "bar"); // <-- this assumes AddressLocator has a constructor accepting Properties
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+
+        //this should fail
+        Hazelcast.newHazelcastInstance(config);
+    }
+
+    @Test
+    public void testPublicIPisTakenFromAddressLocator() {
+        Config config = new Config();
+        config.getNetworkConfig().getAddressLocatorConfig()
+                .setEnabled(true)
+                .setClassname(SimpleAddressLocator.class.getName());
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+        Member localMember = instance.getCluster().getLocalMember();
+
+        assertEquals("1.2.3.4", localMember.getAddress().getHost());
+    }
+
+    public static final class AddressLocatorWithStaticProperties implements AddressLocator {
+        public static Properties properties;
+
+        public AddressLocatorWithStaticProperties(Properties properties) {
+            this.properties = properties;
+        }
+
+        @Override
+        public InetSocketAddress getBindAddress() {
+            return new InetSocketAddress("localhost", 0);
+        }
+
+        @Override
+        public InetSocketAddress getPublicAddress() {
+            return new InetSocketAddress("localhost", 0);
+        }
+    }
+
+    public static final class SimpleAddressLocator implements AddressLocator {
```

Comment:
Suggest to use mocking, also `AddressLocator` implementation classes may throw Runtime exceptions, since they will adapt some other API for us.

Reply:
What happens on core side if concrete implementation returns non-routable IP?

Reply:
are we sure that we do not introduce a security vulnerability?

Reply:
I have adapted some tests to use mocking. Also, I've made tests where the implementation throws a `RuntimeException`. In those cases, the instance will fail to start. This seems fitting since the node fill fail to start if there is an exception in the `DefaultAddressPicker` as well.